File: control-sdk-python/src/control_sdk/checkpoint.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from .control_state import ControlState

# Deliberately generic: this module knows nothing about any specific agent's
# Run/Step types (agent-runtime depends on control-sdk, never the other way
# around) — it just needs a run_id and "which step comes next."


def save_checkpoint(run_id: str, next_step_index: int, control_state: ControlState, path: str | Path) -> None:
    """Write down exactly enough to resume later: which run, which step
    comes next, and the control state at the moment we stopped."""
    data = {
        "run_id": run_id,
        "next_step_index": next_step_index,
        "control_state": control_state.state.value,
        "state_version": control_state.state_version,
    }
    Path(path).write_text(json.dumps(data, indent=2))


def load_checkpoint(path: str | Path) -> dict[str, Any]:
    return json.loads(Path(path).read_text())

File: control-sdk-ts/src/checkpoint.ts
// Same contract as control-sdk-python/src/control_sdk/checkpoint.py
//
// Deliberately generic: this module knows nothing about any specific
// agent's Run/Step types (agent code depends on control-sdk, never the
// other way around) — it just needs a runId and "which step comes next."

import { readFileSync, writeFileSync } from "node:fs";
import type { ControlState } from "./controlState.ts";

export interface CheckpointData {
  runId: string;
  nextStepIndex: number;
  controlState: string;
  stateVersion: number;
}

/** Write down exactly enough to resume later: which run, which step comes
 * next, and the control state at the moment we stopped. */
export function saveCheckpoint(
  runId: string,
  nextStepIndex: number,
  controlState: ControlState,
  path: string,
): void {
  const data: CheckpointData = {
    runId,
    nextStepIndex,
    controlState: controlState.state,
    stateVersion: controlState.stateVersion,
  };
  writeFileSync(path, JSON.stringify(data, null, 2));
}

export function loadCheckpoint(path: string): CheckpointData {
  return JSON.parse(readFileSync(path, "utf-8")) as CheckpointData;
}
